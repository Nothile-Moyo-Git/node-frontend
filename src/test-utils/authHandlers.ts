// handlers.ts
import { graphql } from "msw";

export const handlers = [
  graphql.query("PostUserDetailsResponse", (req, res, ctx) => {
    return res(
      ctx.data({
        PostUserDetailsResponse: {
          sessionCreated: "2024-01-01",
          sessionExpires: "2024-12-31",
          success: true,
          user: {
            _id: "1",
            name: "Nothile Moyo",
            email: "nothile@example.com",
            password: "secret",
            confirmPassword: "secret",
            status: "active",
            posts: [],
          },
        },
      }),
    );
  }),
];
