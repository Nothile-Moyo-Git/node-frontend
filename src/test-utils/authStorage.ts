/**
 * Date Created: 05/08/2025
 *
 * Author: Nothile Moyo
 *
 * Description: This is the local storage handler file for my tests
 * It will handle anything that needs to be set in local storage, in this case, authentication
 *
 * All local storage items are deleted when the request is complete
 */

/**
 * @name setMockAuthStorage
 *
 * @description Mocks setting authentication data for a user
 *
 * @type () => void
 */
export const setMockAuthStorage = ({
  token = "mock-token",
  userId = "mock-user-id",
  expiresIn = new Date(Date.now() + 3600000).toISOString(),
} = {}) => {
  localStorage.setItem("token", token);
  localStorage.setItem("userId", userId);
  localStorage.setItem("expiresIn", expiresIn);
};

export const clearAuthStorage = () => {
  localStorage.removeItem("token");
  localStorage.removeItem("userId");
  localStorage.removeItem("expiresIn");
};
