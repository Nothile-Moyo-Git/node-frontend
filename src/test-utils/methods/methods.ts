// Mock our fetch functioality
export const createFetchResponse = (
  data: unknown,
  status = 200,
  ok = true,
): Response => {
  return {
    ok,
    status,
    json: async () => data,
  } as Response;
};
