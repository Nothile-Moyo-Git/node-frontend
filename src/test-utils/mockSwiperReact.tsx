/**
 * Date created : 01/10/2025
 * Author : Nothile Moyo
 *
 * Test routing file, this will create routes which can be used in Jest for testing
 * We use this to be able to render our components
 *
 */

import { ReactElement } from "react";

export const Swiper = (children: ReactElement) => (
  <div data-testid="swiper">{children}</div>
);

export const SwiperSlide = (children: ReactElement) => {
  <div data-testid="swiper-slide">{children}</div>;
};

export default {
  Swiper,
  SwiperSlide,
};
