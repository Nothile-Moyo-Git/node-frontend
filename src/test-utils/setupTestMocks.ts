/**
 * Date created: 30/07/2025
 *
 * Author: Nothile
 *
 * Methods test file, this file defines reusable methods and hooks which can be used across my tests
 * These methods are intended to be used to replicate the official functionality of the app
 *
 */

import { useState } from "react";
import { User } from "../@types";

// Mock useNavigate so it works with jest
export const mockUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUsedNavigate,
}));

// This allows us to be able to mock hooks in Jest otherwise we use the actual hook which triggers errors
jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

// Cast since we're using a ts file and we want to manipulate this
const mockedUseState = useState as jest.Mock;

const setMockState = jest.fn();

// Default mock implementation of usestate
mockedUseState.mockImplementation((init) => [init, setMockState]);

// Utility function to update mocked isLoading in tests
export const setAppStateMock = (
  isLoading: boolean,
  loadingError: boolean,
  user: User | undefined,
  sessionExpiryDate: string | undefined,
  sessionCreationDate: string | undefined,
) => {
  // Call the state in the same order but ensure that our initial values aren't set multiple times
  let hasSetLoading = false;
  let hasSetUser = false;
  const hasSetExpiry = false;

  mockedUseState.mockImplementation((init) => {
    if (typeof init === "boolean" && !hasSetLoading) {
      hasSetLoading = true;
      return [isLoading, setMockState];
    }
    if (typeof init === "boolean") {
      return [loadingError, setMockState];
    }
    if (init === undefined && !hasSetUser) {
      hasSetUser = true;
      return [user, setMockState];
    }
    /* if (typeof init === "string" && !hasSetExpiry) {
      hasSetExpiry = true;
      return [sessionExpiryDate, setMockState];
    }
    if (typeof init === "string") {
      return [sessionCreationDate, setMockState];
    } */
    return [init, setMockState];
  });
};
