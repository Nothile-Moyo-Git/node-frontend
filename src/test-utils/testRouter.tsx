/**
 * Date created : 30/07/2025
 * Author : Nothile Moyo
 *
 * Test routing file, this will create routes which can be used in Jest for testing
 * We use this to be able to render our components
 *
 */
import { ReactNode } from "react";
import { render } from "@testing-library/react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";

export function renderWithRouter(ui: ReactNode, { route = "/" } = {}, context) {
  const router = createMemoryRouter(
    [
      {
        path: "/",
        element: ui,
      },
    ],
    { initialEntries: [route] },
  );

  return render(<RouterProvider router={router} />);
}
