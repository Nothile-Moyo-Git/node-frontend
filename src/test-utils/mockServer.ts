import { setupServer } from "msw/node";
import { handlers } from "./authHandlers";

export const server = setupServer(...handlers);
