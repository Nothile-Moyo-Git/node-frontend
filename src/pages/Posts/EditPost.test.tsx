/**
 * Author: Nothile Moyo
 * Date created: 25/09/2025
 *
 * @description: This is the test file for the Edit Post component
 * We authenticate the user, and if valid, we then show a list of all posts on the page
 */

import "@testing-library/jest-dom";
import { server } from "../../test-utils/mockServer";
import {
  clearAuthStorage,
  setMockAuthStorage,
} from "../../test-utils/authStorage";
import { createFetchResponse } from "../../test-utils/methods/methods";
import {
  mockContext,
  mockPost,
  mockUser,
} from "../../test-utils/mocks/objects";
import { renderWithContext } from "../../test-utils/testRouter";
import { EditPost } from "./EditPost";
import { screen } from "@testing-library/react";

// Setup mocks and environment
beforeAll(() => {
  server.listen();
});

// Clear our tests and get mock our fetch so we get the correct ordering
beforeEach(() => {
  setMockAuthStorage();
});

// Cleanup mocks and environment
afterEach(() => {
  clearAuthStorage();
});

// End server polling when tests finish
afterAll(() => {
  server.close();
});

describe("Edit Post Component", () => {
  // Handle the api requests
  it("Matches the screenshot", async () => {
    global.fetch = jest
      .fn()
      .mockResolvedValueOnce(
        createFetchResponse({
          data: {
            GetAndValidatePostResponse: {
              success: false,
              message: "200: Request successful",
              post: mockPost,
              isUserValidated: true,
            },
          },
        }),
      )
      .mockResolvedValueOnce(
        createFetchResponse({
          data: {
            PostEditPostResponse: {
              post: mockPost,
              user: mockUser,
              status: true,
              message: "200: Request successful",
              fileValidProps: {
                fileName: mockPost.fileName,
                imageUrl: mockPost.imageUrl,
                isImageUrlValid: true,
                isFileSizeValid: true,
                isFileTypeValid: true,
                isFileValid: true,
              },
              isContentValid: true,
              isTitleValid: true,
              isPostCreator: true,
            },
          },
        }),
      );

    // Render our component with routing and the context so we have authentication
    renderWithContext(
      <EditPost />,
      { route: `/edit-post/${mockPost._id}` },
      mockContext,
    );

    // Make sure we have our edit post
    const editPostComponent = screen.getByTestId("test-id-edit-post");
    expect(editPostComponent).toBeVisible();
    expect(editPostComponent).toMatchSnapshot();
  });
});
