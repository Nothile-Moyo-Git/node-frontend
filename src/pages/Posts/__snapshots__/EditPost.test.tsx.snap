// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Post Component Completely loads the page and updates the content successfully 1`] = `
<section
  class="editPost"
  data-testid="test-id-edit-post"
>
  <form
    class="form  false"
  >
    <h2
      class="title false"
    >
      Edit Post : 2B
    </h2>
    <div
      class="field "
    >
      <label
        class="label false"
        for="title"
        id="titleLabel"
      >
        Title*
      </label>
      <input
        aria-labelledby="titleLabel"
        class="input  false"
        name="title"
        placeholder="Enter your title here"
        required=""
        type="string"
        value="2B"
      />
    </div>
    <div
      class="field "
    >
      <div
        class="loadingWrapper"
        data-testid="test-id-loading-spinner"
      >
        <span
          class="loadingSpinner"
        />
        <p
          class="loadingText"
        >
          Loading, please wait...
        </p>
      </div>
    </div>
    <div
      class="field "
    >
      <label
        class="label false"
        for="imageUrl"
        id="imageUrlLabel"
      >
        Previous image: 2B.png
      </label>
      <div
        class="imagePreview"
        style="background-image: url(test-file-stub); background-size: contain; background-position: left;"
      />
    </div>
    <div
      class="field "
    >
      <label
        class="label false"
        for="content"
        id="contentLabel"
      >
        Content*
      </label>
      <textarea
        aria-labelledby="contentLabel"
        class="textarea  false"
        name="content"
        placeholder="Please enter your content"
        required=""
        rows="3"
      >
        YoRHa No. 2 Type B,[a] commonly known as 2B,[b] is a fictional android from the 2017 video game Nier: Automata, a spin-off of the Drakengard series developed by PlatinumGames and published by Square Enix. One of the game's three protagonists, 2B is a soldier for YoRHa, an android task force fighting a proxy war with alien-created Machine Lifeforms. Her story arc focuses on her backstory within YoRHa, and her relationship with her partner 9S, a reconnaissance android. She is also featured in related media, such as the anime Nier: Automata Ver1.1a.
      </textarea>
    </div>
    <button
      class="button button__primary"
      data-testid="test-id-edit-post-submit-button"
      type="submit"
    >
      Submit
    </button>
  </form>
</section>
`;

exports[`Edit Post Component Matches the snapshot 1`] = `
<section
  class="editPost"
  data-testid="test-id-edit-post"
>
  <form
    class="form  false"
  >
    <h2
      class="title false"
    >
      Edit Post : 2B
    </h2>
    <div
      class="field "
    >
      <label
        class="label false"
        for="title"
        id="titleLabel"
      >
        Title*
      </label>
      <input
        aria-labelledby="titleLabel"
        class="input  false"
        name="title"
        placeholder="Enter your title here"
        required=""
        type="string"
        value="2B"
      />
    </div>
    <div
      class="field "
    >
      <div
        class="loadingWrapper"
        data-testid="test-id-loading-spinner"
      >
        <span
          class="loadingSpinner"
        />
        <p
          class="loadingText"
        >
          Loading, please wait...
        </p>
      </div>
    </div>
    <div
      class="field "
    >
      <label
        class="label false"
        for="imageUrl"
        id="imageUrlLabel"
      >
        Previous image: 2B.png
      </label>
      <div
        class="imagePreview"
        style="background-image: url(test-file-stub); background-size: contain; background-position: left;"
      />
    </div>
    <div
      class="field "
    >
      <label
        class="label false"
        for="content"
        id="contentLabel"
      >
        Content*
      </label>
      <textarea
        aria-labelledby="contentLabel"
        class="textarea  false"
        name="content"
        placeholder="Please enter your content"
        required=""
        rows="3"
      >
        YoRHa No. 2 Type B,[a] commonly known as 2B,[b] is a fictional android from the 2017 video game Nier: Automata, a spin-off of the Drakengard series developed by PlatinumGames and published by Square Enix. One of the game's three protagonists, 2B is a soldier for YoRHa, an android task force fighting a proxy war with alien-created Machine Lifeforms. Her story arc focuses on her backstory within YoRHa, and her relationship with her partner 9S, a reconnaissance android. She is also featured in related media, such as the anime Nier: Automata Ver1.1a.
      </textarea>
    </div>
    <button
      class="button button__primary"
      data-testid="test-id-edit-post-submit-button"
      type="submit"
    >
      Submit
    </button>
  </form>
</section>
`;
